contracts/AnyswapV4CallProxy.sol

```diff
@@ -151,20 +200,25 @@ contract AnyCallProxy is Whitelistable {
         address[] memory callbacks,
         uint256[] memory nonces,
         uint256 fromChainID
-    ) external onlyMPC lock {
+    ) external onlyMPC lock bill(from) {
```

Comment:
Billing is done as a modifier, and tracks the execution cost of the entire function (without the onlyMPC / reentrancy locks)

Reply:
Nice design, top up in advance, billing based on gas usage.

Reply:
Do you think it resonable if we offer different price to different dapp?

Reply:
> Nice design, top up in advance, billing based on gas usage.

@gaozhengxin Thanks !! I think the billing still needs a bit more refining (like including a fixed overhead to make it marginally profitable to AnySwap). And thinking about it, the `bill` modifier should come before the `onlyMPC` and `lock` modifiers so that they're apart of the billing.

> Do you think it resonable if we offer different prices to different dapps?

Hmm ... :thinking:, yeah I think it is. There should be some profit for AnySwap, especially since there is the infrastructure behind the scenes you guys are running, and this is a service. Any thoughts on how you'd want to implement? Do you think accepting only the network native token is fine, or do you want to maybe accept ERC20 (DAI/USDC/etc..) <--- this might be best since its a stablecoin and won't have to worry about volatility. Do you want to do per-tx (.50 USDC per call per dapp)  / ($250 /month per dapp).

In anycase, with a per dapp surcharge, we'll probably want a mapping like `mapping(bytes32 => Pricing) pricing;`, where `bytes32` is a key, probably `keccak256({INSERT_DAPP_NAME})`, and the Pricing struct has whatever data we need

Reply:
`bill onlyMPC lock` ✔

We'd like to take native token to cover gas usage and stablecoin for profit. I think fee-per-call is good enough for us. I think we should do billing against target contract, and let funders pay. And we prefer call refund by MPC and receive funds in a wallet address.
```
abstract contract Billable is Whitelistable {
    struct Fund {
        uint256 nativeToken;
        address token;
        uint256 amount;
    }
    mapping(address => address) public funder;
    mapping(address => Fund) public funds;
    mapping(address => Fund) public pricing;

    modifier bill(address target) {
        // cost = expense + pricing[target]
        funds[funder[target]] -= int256(cost);
    }

    function nativeRefund(address receiver) onlyMPC {
        // send expense to receiver
    }

    function erc20Refund(address token, address receiver) onlyMPC {
        // send token balance to receiver
    }
}
```